30413/Andreea Alexandra Stan/Lab09/lab9_mandatory/main.c

```diff
@@ -0,0 +1,139 @@
+#include <stdio.h>
+#include <stdlib.h>
+#include <limits.h>
+#include <math.h>
+
+FILE *openAndCheck(char *filePath, char *mode)
+{
+    FILE *f = fopen(filePath, mode);
+    if (f==NULL)
+    {
+        printf("Cannot open the file");
+        exit(0);
+    }
+    return f;
+}
+
+
+//i divide the array in two parts
+//the max sum can be either in the left array, in the right one or it can contain elements from both parts
+
+
+//function to get the maximum from three number
+int maxim(int a, int b, int c)
+{
+    if(a>=b && a>=c)
+        return a;
+    else if (b>=a && b>=c)
+        return b;
+    else
+        return c;
+
+}
+
+
+//function to get the maximum sum of consecutive elements of the subarray, including the one in the middle
+int max_midsum(int a[],int left, int mid, int right)
+{
+    int i = mid+1;
+    int j = mid;
+    int sum_left=0,sum_right=0;
+
+    int maximl=INT_MIN;
+    int maximr=INT_MIN;
+
+    //start from the middle element (including it) and get the maximum of consecutive elements on its left side
+    while(j >= left)
+    {
+        sum_left=sum_left+a[j];
+
+        if(sum_left>maximl)
+            maximl =sum_left;
+        j--;
+    }
+
+    //repeat the same process for the right side(this time we don't include the middle element):
+    while(i<=right)
+    {
+        sum_right=sum_right+a[i];
+
+        if(sum_right>maximr)
+            maximr =sum_right;
+
+        i++;
+    }
+
+    //then we add the left maximum sum and the right maximum sum
+    return maximl+maximr;
+
+
+}
+
+int maximl,maximr,maxim_intersect;
+
+//this is the part where i divide the problem into subproblems
+//first i divide the array into two parts
+//the max sum can be either in the left subarray, in the right one or it can contain elements from both parts
+//we check the third case with the max_midsum function
+
+//with this function I check the three possible cases, and return the maximal one
+
+int divide_et_impera(int a[],int left, int right)
+{
+
+    if(left < right)
+    {
+        int mid = (left + right) / 2;
+
+        maximl=divide_et_impera(a,left, mid);
+        maximr=divide_et_impera(a, mid + 1, right);
+        maxim_intersect=max_midsum(a,left, mid, right);
+
+        return maxim(maximl,maximr,maxim_intersect);
+    }
+
+
+}
```

Comment:
Aici am un warning "control reaches end of non-void function" si nu prea inteleg de ce.

Reply:
Daca programul nu intra in "if" atunci ce returneaza? Trebuie sa mai pui un return.